### DISCUSSION TITLE
Migration of code for pint < 0.20 to newer version

### DISCUSSION THREAD

**Discussion by fxcoudert** [hgrecco/pint#2097] [Q&A] [answered]
Hi,

I'm trying to migrate some code that worked only on `pint<0.20.0` (context is given in https://github.com/materialsproject/MPContribs/issues/1871). It was written like this:

```
from pint import UnitRegistry
from pint.unit import UnitDefinition
from pint.converters import ScaleConverter
from pint.errors import DimensionalityError

ureg = UnitRegistry(
    autoconvert_offset_to_baseunit=True,
    preprocessors=[
        lambda s: s.replace("%%", " permille "),
        lambda s: s.replace("%", " percent "),
    ],
)
ureg.define(UnitDefinition("percent", "%", (), ScaleConverter(0.01)))
ureg.define(UnitDefinition("permille", "%%", (), ScaleConverter(0.001)))
ureg.define(UnitDefinition("ppm", "ppm", (), ScaleConverter(1e-6)))
ureg.define(UnitDefinition("ppb", "ppb", (), ScaleConverter(1e-9)))
ureg.define("atom = 1")
ureg.define("bohr_magneton = e * hbar / (2 * m_e) = µᵇ = µ_B = mu_B")
ureg.define("electron_mass = 9.1093837015e-31 kg = mₑ = m_e")
```

This is intialization-level code, after which there is code for unit conversion, which I believe will not be a problem.

The code above fails with recent pint, because it cannot find `UnitDefinition` and `ScaleConverter`. However, I was looking at the list of breaking changes in the pint ChangeLog, but I do not see anything that would cover this.

Could someone give advice on how to update this code?

**Comment by andrewgsavage** [comment]
it looks like you got this working in the linked PR, is this still needed?

**Reply by fxcoudert** [reply]
I'd still appreciate some link to documentation about why this change is needed / where it was introduced.

**Reply by andrewgsavage** [reply, marked as answer]
UnitDefinition and ScaleConverter are now in pint.facets.plain after the parser was split into its own module, flexparser

I think they arent documented with examples as its not expected users will need to use them
